torch/distributed/_tensor/api.py

```diff
@@ -213,6 +213,28 @@ def __repr__(self):
         # TODO: consider all_gather the local tensors for better debugging
         return f"DTensor(local_tensor={self._local_tensor}, device_mesh={self._spec.mesh}, placements={self._spec.placements})"
 
+    def __tensor_flatten__(self):
```

Comment:
I am not 100% sold on this being a protocol/API we want to exposed versus just doing this behavior for DTensor only. Thoughts?

Reply:
I think we can definitely do this behavior only for DTensor by special casing DTensor fakification in dynamo, but I think for other tensor subclasses this might be useful as long as they defined the similar interface, i.e. @bdhirsh is also trying to enable `parseSemiStructuredTensor` which also needs to define `__tensor_flatten__` and `__tensor_unflatten__` https://fburl.com/zz802okh, so I feel exposing this protocol might makes more uniform?

Reply:
I'm definitely pro having a generalized APIs; there are more subclasses than just DTensor. If we need to bikeshed the protocol, we can, but I think we should have *something*.